nexus/db-queries/src/db/pool.rs

```diff
@@ -3,108 +3,144 @@
-        LoggingErrorSink { log }
-    }
+fn make_postgres_connector() -> qorb::backend::SharedConnector<QorbConnection> {
+    // Create postgres connections.
+    //
+    // We're currently relying on the DieselPgConnector doing the following:
+    // - Disallowing full table scans in its implementation of "on_acquire"
+    // - Creating async_bb8_diesel connections that also wrap DTraceConnections.
+    let user = "root";
+    let db = "omicron";
+    let args = Some("sslmode=disable");
+    Arc::new(DieselPgConnector::new(user, db, args))
 }
 
-impl bb8::ErrorSink<ConnectionError> for LoggingErrorSink {
-    fn sink(&self, error: ConnectionError) {
-        error!(
-            &self.log,
-            "database connection error";
-            "error_message" => #%error
-        );
+impl Pool {
+    /// Creates a new qorb-backed connection pool to the database.
+    ///
+    /// Creating this pool does not necessarily wait for connections to become
+    /// available, as backends may shift over time.
+    pub fn new_qorb(bootstrap_dns: Vec<SocketAddr>) -> Self {
+        // Make sure diesel-dtrace's USDT probes are enabled.
+        usdt::register_probes().expect("Failed to register USDT DTrace probes");
+
+        let resolver = make_dns_resolver(bootstrap_dns);
+        let connector = make_postgres_connector();
+
+        let policy = Policy::default();
+        Pool { inner: qorb::pool::Pool::new(resolver, connector, policy) }
+    }
+
+    /// Creates a new qorb-backed connection pool to a single instance of the
+    /// database.
+    ///
+    /// This is intended for tests that want to skip DNS resolution, relying
+    /// on a single instance of the database.
+    ///
+    /// In production, [Self::new_qorb] should be preferred.
+    pub fn new_qorb_single_host(db_config: &DbConfig) -> Self {
+        // Make sure diesel-dtrace's USDT probes are enabled.
+        usdt::register_probes().expect("Failed to register USDT DTrace probes");
+
+        let resolver = make_single_host_resolver(db_config);
+        let connector = make_postgres_connector();
+
+        let policy = Policy::default();
+        Pool { inner: qorb::pool::Pool::new(resolver, connector, policy) }
+    }
+
+    /// Creates a new qorb-backed connection pool which returns an error
+    /// if claims are not quickly available.
+    ///
+    /// This is intended for test-only usage.
+    pub fn new_qorb_single_host_failfast(db_config: &DbConfig) -> Self {
+        // Make sure diesel-dtrace's USDT probes are enabled.
+        usdt::register_probes().expect("Failed to register USDT DTrace probes");
+
+        let resolver = make_single_host_resolver(db_config);
+        let connector = make_postgres_connector();
+
+        let policy = Policy {
+            claim_timeout: tokio::time::Duration::from_millis(1),
+            ..Default::default()
+        };
+        Pool { inner: qorb::pool::Pool::new(resolver, connector, policy) }
```

Comment:
perhaps these should be gated behind a `test_util` feature or something, which can be enabled by dev-dependencies on this crate?

Reply:
Good idea -- this is a little harder to do on `new_single_host`, because of the current clients and nexus configs, but I can definitely do this for the failfast one at least